### ISSUE TITLE
Comparison of dynamic types of elements and conditions

### ISSUE THREAD

**Issue by msandre** [KratosMultiphysics/Kratos#1427]
This is more of a question to the developers of model_part_io. There is a common need to compare the types of elements and conditions in kratos, in hdf5 app and in `ModelPartIO::WriteElements`. It makes sense have a common implementation in the core for this.

In `ModelPartIO::WriteElements` one compares types with

`IsSameElement<decltype(*(rThisElements.begin())), decltype(*(it_comp->second))>::value`

and geometries with

`element_type == (it_comp->second)->GetGeometry().GetGeometryType()`.

In the hdf5 it uses 

`typeid(lhs) == typeid(rhs)`

and compares nnodes and dim for the geometry. I agree `element_type == (it_comp->second)->GetGeometry().GetGeometryType()` is the better solution for geometry, but as far as I understand, `IsSameElement<decltype(*(rThisElements.begin())), decltype(*(it_comp->second))>::value` compares the static types when what we want is the dynamic types. Should it not use typeid?

If this is the same I will add the utility to the core and update model_part_io and hdf5 to use the same implementation.

Thanks

**Comment by loumalouomega** [comment]
I remember that in that moment we implemented that way because the standard typeid was failing and we decided this one instead

**Comment by roigcarlo** [comment]
The specific implementation you see in model_part_io.h is done that way to avoid warnings as the Component against we are comparing is marked as const. IsSameElement is just an alias for `std::is_same` which strips off the `volatile` and `const` qualifiers from the type.

Not sure what you mean by `dynamic types`. If you mean derived ones as @loumalouomega says you cannot use `typeid`, you need to evaluate the type of the expression with `decltype`, unless you have the instance of the type you want to evaluate with `typeid` ( which is not the case there, as we have the pointer).

TL;DR, its basically the same but we do it that way to use directly the pointers and avoid warnings

**Comment by msandre** [comment]
hmmm, then I'm quite confused by this code https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/sources/model_part_io.cpp#L269

I made a very simplified test to try to understand what is happening here. It may easily be that I am missing something, but I think this just compares the static types (i.e., is Element == Element) in which case it could be removed. But then imagine if rThisElements has two element types (say FluidElement and SolidElement) both with the same geometry type. Let rThisElements.begin() return an iterator pointing to a FluidElement and KratosComponents<Element>::GetComponents().begin() return an iterator pointing to a SolidElement. Then as I understand, it will think the first element in the container is SolidElement and start a block `Begin Elements "SolidElement"`. 

Sorry if I'm not getting it, but I don't see how this code can work in such cases. By dynamic type I meant the polymorphic type of the object at runtime.

**Comment by roigcarlo** [comment]
That code compares the polymorphic type. Both are elements but are not the same kind of element. I don't know exactly the example in which the comparison is needed. @loumalouomega can help you there.

**Comment by loumalouomega** [comment]
I don't understand, an element is an  element, they can share the geometry, but there is only one type of geometry, we can create a test to check it properly works independently of the polymorphic type

**Comment by roigcarlo** [comment]
The think is, for example in the write function we do need the element type
to print its name, and we cannot known it unless we compare it with the
component. About the geometries i do not know sorry.

**Comment by msandre** [comment]
I'm setting up a test more rigorously now to understand it better. Will post it when finished.

**Comment by msandre** [comment]
Here I modified test_model_part_io.cpp and used a fluid element to illustrate the point.

```
KRATOS_TEST_CASE_IN_SUITE(ModelPartIOWriteModelPart2, KratosCheckWriteElementsTestSuite)
{
    std::cout << std::endl;
    for(auto it = KratosComponents<Element>::GetComponents().begin();
        it != KratosComponents<Element>::GetComponents().end(); it++)
    {
        std::cout << it->first << std::endl;
    }

    // Create a model part to write
    ModelPart main_model_part("MainModelPart");
    main_model_part.SetBufferSize(1);
    Properties::Pointer p_properties_1(new Properties(1));
    main_model_part.AddProperties(p_properties_1);

    main_model_part.CreateNewNode(1, 0.0, 0.0, 0.0);
    main_model_part.CreateNewNode(2, 1.0, 0.0, 0.0);
    main_model_part.CreateNewNode(3, 1.0, 1.0, 0.0);

    std::vector<ModelPart::IndexType> elem_nodes_1 = {1, 2, 3};
    main_model_part.CreateNewElement("VMS2D3N", 1, elem_nodes_1, p_properties_1);

    // Create the output .mdpa file
    std::string output_file_name = "main_model_part_output";
    std::fstream output_file;
    output_file.open(output_file_name + ".mdpa", std::fstream::out);
    output_file.close();

    // Fill the output .mdpa file
    ModelPartIO model_part_io_write(output_file_name, IO::WRITE);
    model_part_io_write.WriteModelPart(main_model_part);
}
``` 

Output:

```
BinghamFractionalStep2D
BinghamFractionalStep3D
BinghamFractionalStepDiscontinuous2D
...
```

main_model_part_output.mdpa:

```
Begin Properties 1

End Properties

Begin Nodes
	1	0	0	0
	2	1	0	0
	3	1	1	0
End Nodes

Begin Elements	BinghamFractionalStep2D
	1	1	1	2	3	
End Elements
```

I need this functionality so I can work on a solution, but I think the current implementation is all compile time checking and so cannot catch the correct element type. As far as I know, we have to go with typeid.

**Comment by roigcarlo** [comment]
Now its me who does not understand it... 
So basically the problem is that IsSameElement is returning true for let's say "BinghamFractionalStep2D" and "VMS2D3N".

If this is the case its a bug, and indeed we have to fix it. I will also take a look at it

**Comment by msandre** [comment]
yes, it should return true for any element, in this case the first one in the kratos components is BinghamFractionalStep2D which has the same geometry type as VMS2D3N so it takes this one.

**Comment by roigcarlo** [comment]
@msandre Basically what you said in the 1st post was right. Seems that I messed up the code while fixing a warning since there were not tests for this and I didn't understand what exactly the code was doing.

I made the fixes in https://github.com/KratosMultiphysics/Kratos/tree/core/fix-modelpart-io. Currently I am finishing the tests for every case
 * Same geom + Different Element
 * Different Element + Same Geom
 * Cases above with repetitions and mixing cases
So this does not happen again...

Once I have the checks for that I'll upload.

So to answer the first question
> If this is the same I will add the utility to the core and update model_part_io and hdf5 to use the same implementation.

Yes it is the same (or should have been), so +1 for the unified implementation.

**Comment by msandre** [comment]
@roigcarlo ok thanks, I added the utility in pr #1434.

**Comment by loumalouomega** [comment]
Thanks @msandre to solve the chaos of this situation

**Comment by roigcarlo** [comment]
Fixed in #1437
